source/common/upstream/logical_host.h

```diff
@@ -0,0 +1,98 @@
+#pragma once
+
+#include "common/upstream/upstream_impl.h"
+
+namespace Envoy {
+namespace Upstream {
+
+/**
+ * A host implementation that can have its address changed in order to create temporal "real"
+ * hosts.
+ */
+class LogicalHost : public HostImpl {
+public:
+  LogicalHost(const ClusterInfoConstSharedPtr& cluster, const std::string& hostname,
+              const Network::Address::InstanceConstSharedPtr& address,
+              const envoy::api::v2::endpoint::LocalityLbEndpoints& locality_lb_endpoint,
+              const envoy::api::v2::endpoint::LbEndpoint& lb_endpoint)
+      : HostImpl(cluster, hostname, address, lb_endpoint.metadata(),
+                 lb_endpoint.load_balancing_weight().value(), locality_lb_endpoint.locality(),
+                 lb_endpoint.endpoint().health_check_config(), locality_lb_endpoint.priority(),
+                 lb_endpoint.health_status()) {}
+
+  // Set the new address. Updates are typically rare so a R/W lock is used to address updates.
+  // Note that the health check address update requires no lock to be held since it is only
+  // used on the main thread.
```

Comment:
Do we have any kind of assert that ensures that this is true? Seems like a thing that could cause subtle issues if that ever changes. Or maybe we could rely on TSAN to catch that?

Reply:
Let me just add locking to be on the safe side. I don't think it's worth it not to now that you mention it.